controller/src/main/java/io/pravega/controller/store/stream/tables/SegmentIndexRecord.java

```diff
@@ -21,26 +21,45 @@
  */
 @Data
 public class SegmentIndexRecord {
-    private static final int INDEX_RECORD_SIZE = Integer.BYTES;
+    // The first (integer) record of this table contains the starting segment number for this stream. All the segments
+    // in this stream will contain ids starting from the the value stored in the first record of this table.
+    static final int STARTING_SEGMENT_NUMBER_POSITION = 0;
+    static final int INDEX_RECORD_SIZE = Integer.BYTES;
     private final int segmentNumber;
     private final int segmentOffset;
+    private final int startingSegmentNumber;
 
     public static Optional<SegmentIndexRecord> readRecord(final byte[] indexTable, final int segmentNumber) {
-        if ((segmentNumber + 1) * INDEX_RECORD_SIZE > indexTable.length || segmentNumber < 0) {
+        // To get the appropriate index for segmentNumber, we need to subtract the starting segment number for this
+        // stream and add 1 to take into account the storage space of the starting segment number value (first element).
+        final int actualSegmentIndex = fromSegmentNumberToTableIndex(indexTable, segmentNumber);
+
+        // Note that the segment index has now an extra header integer to store the starting segment number.
+        if ((actualSegmentIndex + 1) * INDEX_RECORD_SIZE > indexTable.length || actualSegmentIndex < 0) {
             return Optional.empty();
         } else {
-            return Optional.of(parse(indexTable, segmentNumber));
+            return Optional.of(parse(indexTable, actualSegmentIndex));
         }
     }
 
     public static Optional<SegmentIndexRecord> readLatestRecord(final byte[] indexTable) {
-        return readRecord(indexTable, indexTable.length / INDEX_RECORD_SIZE - 1);
+        final int startingSegmentNumber = BitConverter.readInt(indexTable, STARTING_SEGMENT_NUMBER_POSITION);
+        return readRecord(indexTable, startingSegmentNumber + indexTable.length / INDEX_RECORD_SIZE - 2);
```

Comment:
instead of doing `-2` here i would have ```(indexTable.length - STARTING_SEGMENT_NUMBER_RECORD_SIZE) / INDEX_RECORD_SIZE - 1```

Reply:
I think that the proposed change is not feasible. The root cause is that this PR breaks the coupling between segment index position and segment number. That is, `SegmentIndexRecord.readRecord()` expects _segment numbers_ that will be then converted into the appropriate positions to access the segment index. However, your proposal is directly providing _a segment index position_ as input to `SegmentIndexRecord.readRecord()`. This does not work with the new approach. Instead, `readLatestRecord` needs to infer the last segment number for this stream so that `SegmentIndexRecord.readRecord()` can access to the appropriate index position. In other words, `startingSegmentNumber` is not `STARTING_SEGMENT_NUMBER_RECORD_SIZE`, but the number stored in the first position of the index that represents the stream starting segment number. As an example, this means that if the segment index has `11` positions (`1` position for the starting segment number + `10` indexed segments) and the starting segment number for this stream is `90`, `readLatestRecord` will execute `readRecord(99)` (last segment number for this stream, which differs from its position in the index).

Reply:
`readRecord` takes segment number that it needs to read. 
```
latest segment number = startingSegmentNumber + `totalNumberOfRecordsInIndex` - 1
```
prior to this PR the startingSegmentNumber was always `0` and 
```totalNumberOfRecordsInIndex = length/PER_RECORD_SIZE```

However, now that we have put `STARTING_SEGMENT_NUMBER` which takes up STARTING_SEGMENT_NUMBER_RECORD_SIZE` space in the length our computation should change accordingly:

To calculate `totalNumberOfRecords`, we need to deduct `STARTING_SEGMENT_NUMBER_RECORD_SIZE` from total length and then divide it by per record size. 
```totalNumberOfRecords = (length - STARTING_SEGMENT_NUMBER_RECORD_SIZE) / PER_RECORD_SIZE```

The reason you could get away with deducting `2` was because `STARTING_SEGMENT_NUMBER_RECORD_SIZE` is same as `PER_RECORD_SIZE` which is 4 bytes (Integer.BYTES)

Reply:
So you probably misunderstood my comment. I am not asking you to deduct startingSegmentNumber but number of bytes taken to store startingSegmentNumber. 
Its only incidental that we use same number of bytes to store an index record and the starting segment number. But we should make the computation more explicit here for future maintainability of the code.

Reply:
Thanks, I understood your point and fixed it.